text/0000-dynamic-events.md

```diff
@@ -0,0 +1,136 @@
+
+
+## Detailed design
+
+A new directive `lwc:on` will be introduced that can be used to add a collection of event listeners whose name may not be known while authoring the component.
+
+### Structure
+
+The `lwc:on` directive would accept an object. For each property of the object, it would add a event listener which would listen for the event specified by the property's key and handle it using the property's value bound to the owner component.
+
+### Caching
+
+#### For static components, i.e. components declared in template directly using its selector
+Since it is uncommon for event listeners to change after the start of Owner component's intial render, We can cache them to improve performance. Note that this is same as how `onevent` on template works currently. For consumers, the implication of this would be that any changes made to the object passed to `lwc:on` after the first `connectedCallback` of owner component would cause no effect.
+
+#### For dynamic components, i.e. components created using `lwc:component` directive
+Since it is uncommon for event listeners to change if the constructor passed to `lwc:is` doesn't change, We can skip patching of event listeners if the element doesn't change. For consumers, the implication of this would be that after the first `connectedCallback` of owner component, any changes made to the object passed to `lwc:on` would cause no effect until the constructor passed to `lwc:is` itself is changed.
+
+
+### Overriding
+
+```js
+// x/myComponent.js
+export default class MyComponent extends LightningElement {
+
+    childEventHandlers = {
+        foo: function (event) {console.log('lwc:on');} ,
+    };
+
+    fooHandler(){
+        console.log('onfoo');
+    }
+
+}
+```
+
+```html
+ <!-- x/myComponent.html -->
+<template>
+  <x-child onfoo={fooHandler} lwc:on={childEventHandlers}></x-child>
+</template>
+```
+
+`lwc:on` will always be applied last. That means it will take precedence over whatever event listeners are declared in the template directly. In the above example, `x-child`, only one listener for event `foo` would be added and `childEventHandlers.foo` would be used for event handler .
+Additionally, there can be only one `lwc:on` directive on a element 
+
+### Event names
+
+The keys of object passed to `lwc:on` should conform to the requirement set by DOM Event Specification. There would be no other constraint on the object's keys.
+
+## Drawbacks
+
+Why should we *not* do this? Please consider:
+
+- implementation cost, both in term of code size and complexity
+- whether the proposed feature can be implemented in user space
+- the impact on teaching people Lightning Web Components
+- integration of this feature with other existing and planned features
+- cost of migrating existing Lightning Web Components applications (is it a breaking change?)
+
+There are tradeoffs to choosing any path. Attempt to identify them here.
+
+
+To add : Static Analyzability.

```

Comment:
For Static Analysis since it's a dynamic component we would run into similar issues as we do with`lwc:component` in that we cannot easily search for what components these create in the client code.